lib/publiccloud/instance.pm

```diff
@@ -119,6 +122,90 @@ sub retry_ssh_command {
     die "Waiting for Godot: " . $cmd;
 }
 
+# Auxilliary function to prepare the ssh command that runs any command on the PC instance
+sub _prepare_ssh_cmd {
+    my ($self, %args) = @_;
+    die('No command defined') unless ($args{cmd});
+    $args{ssh_opts} //= $self->ssh_opts() . " -i '" . $self->ssh_key . "'";
+    $args{username} //= $self->username();
+    $args{timeout} //= SSH_TIMEOUT;
+
+    my $cmd = $args{cmd};
+    unless ($args{no_quote}) {
+        $cmd =~ s/'/\'/g;    # Espace ' character
+        $cmd = "\$'$cmd'";
+    }
+
+    my $ssh_cmd = sprintf('ssh %s "%s@%s" -- %s', $args{ssh_opts}, $args{username}, $self->public_ip, $cmd);
+    #$ssh_cmd = "timeout $args{timeout} $ssh_cmd" if ($args{timeout} > 0);
+    return $ssh_cmd;
+}
+
+=head2 ssh_script_run
+
+    ssh_script_run($cmd [, timeout => $timeout] [, fail_message => $fail_message] [,quiet => $quiet] [,ssh_opts => $ssh_opts] [,username => $username])
+
+Runs a command C<cmd> via ssh on the publiccloud instance and returns the return code.
+=cut
+sub ssh_script_run {
+    my $self = shift;
+    my %args = testapi::compat_args({cmd => undef}, ['cmd'], @_);
+    my $ssh_cmd = $self->_prepare_ssh_cmd(%args);
+    delete($args{cmd});
+    delete($args{ssh_opts});
+    delete($args{username});
```

Comment:
What about `$self->instance->username`?

Reply:
I guess the purpose was to being able to pass an arbitrary username (e.g. `root`) but I'm not aware of it being anywhere used.
In the end it's best to stay close to the original `run_ssh_command` implementation, that's why this is still there.

Reply:
it is very useful flow which I have used several times . For example SLEM pubcloud images has hardcoded user `suse` and current master allowing simple fire test with  `PUBLIC_CLOUD_USER` to perfectly handle this